### DISCUSSION TITLE
How to perform an action after a ContentItem shape is displyed?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#13484] [Q&A]
I am trying to perform an action after a ContentItem has been displayed.

I created a custom ShapeTable like this

```
public class ViewingShapeTable : IShapeTableProvider
{
    public void Discover(ShapeTableBuilder builder)
    {
        builder.Describe("ContentItem")
            .OnDisplayed(context =>
            {
                // perform my action
            });
        
    }
}
```

However, the `OnDisplayed` action is never called when I view a ContentItem as I would expect.

Of cource I registered the shape table like this

```
services.AddScoped<IShapeTableProvider, ViewingShapeTable>();
```

While tracking, the `Discover(ShapeTableBuilder builder)` gets called, but none of the `On...` actions are getting called.

@ns8482e you probably know the reason my action isn't being called.

**Comment by hishamco** [comment]
Probably you need to decorate your `ViewingShapeTable` with `[Feature]` in case this shape runs on a particular feature. I don't see your code, but this happen with my last time :)

**Reply by jtkech** [reply]
The parameter of `.Describe("...")` should be a shape type name (not an alternate), `ContentItem` is not a shape type, `Content` is.

Here an example I'm using.

    public void Discover(ShapeTableBuilder builder) => builder
        .Describe("ListPartDetailAdmin")
        .OnDisplaying(context =>
        {
            if (context.Shape is ListPartViewModel model && model.ListPart.ContentItem.ContentType == Types.LearningPathway)
            {
                model.ContentItems = model.ContentItems
                    .OrderBy(lesson => lesson.As<LessonPart>()
                        ?.PathwaySequenceNumber.Value ?? 0);
            }
        });

**Reply by MikeAlhayek** [reply]
@jtkech i actually tried `Content` too prior posting this question. That did not work either. This shows ContentItem described as a shape too https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Contents/Shapes.cs

**Reply by MikeAlhayek** [reply]
@hishamco I missed your comment. I think as long as that service is registered within a startup that belongs to a feature, you don't need to add feature attribute on the service. In face, the builder method is called, but the shape action isn't being called

**Reply by hishamco** [reply]
> you don't need to add feature attribute on the service

Are you 100% sure :)

**Reply by jtkech** [reply]
Just tried, it works with `Content` but not with `ContentItem` for which would need to create and drive such a shape.

**Reply by MikeAlhayek** [reply]
If the registration was the issue, the Discover method would have never been called. So yes I am sure on that. But not sure why the Shape action isn't being called.

**Reply by MikeAlhayek** [reply]
@jtkech you are right, `Content` not `ContentItem`. However, my content item has a stereotype so the shape is my stereotype not content. 

Thanks for your help!

**Comment by ns8482e** [comment]
`ContentItem` shape used only when you use `contentitem` tag helper.

When you are rendering content item using content display manager you are actually using Content shape and not ContentItem shape
